phoenix-core/src/main/java/org/apache/phoenix/optimize/QueryOptimizer.java

```diff
@@ -258,6 +259,16 @@ public QueryPlan optimize(QueryPlan dataPlan, PhoenixStatement statement, List<?
             throw new RowValueConstructorOffsetNotCoercibleException("No table or index could be coerced to the PK as the offset. Or an uncovered index was attempted");
         }
 
+        if (applicablePlans.get(0) instanceof CursorFetchPlan) {
```

Comment:
I would rather put in a generic query method if possible.  This code seems to imply that the cursor fetch plan is always first followed by several normal execution plans.  This special case handling seems awkward at best is. it possible to generalize this somehow either in initial plan creation ie create them as separate cursor plans or maybe as part of a general plan method?  I probably am missing a bunch of details.

Reply:
Thank you for the review @dbwong !